unixfs/src/dir/builder/iter.rs

```diff
@@ -0,0 +1,416 @@
+                        // we wouldn't need to go through Visited::Post.
+                    }
+                }
+                Visited::Post {
+                    parent_id,
+                    id,
+                    name,
+                    leaves,
+                    ..
+                } => {
+                    // all of our children have now been visited; we should be able to find their
+                    // Cids in the btreemap
+                    let mut collected = self.persisted_cids.remove(&Some(id)).unwrap_or_default();
+
+                    // FIXME: leaves could be drained and reused
+                    collected.extend(leaves);
+
+                    if !self.wrap_with_directory && parent_id.is_none() {
+                        // we aren't supposed to wrap_with_directory, and we are now looking at the
+                        // possibly to be generated root directory.
+
+                        assert_eq!(
+                            collected.len(),
+                            1,
+                            "should not have gone this far with multiple added roots"
+                        );
+
+                        return None;
+                    }
+
+                    let buffer = &mut self.block_buffer;
+
+                    let leaf = match Self::render_directory(&collected, buffer) {
+                        Ok(leaf) => leaf,
+                        Err(e) => return Some(Err(e)),
+                    };
+
+                    self.cid = Some(leaf.link.clone());
+                    self.total_size = leaf.total_size;
+
+                    // this reuse strategy is probably good enough
+                    collected.clear();
+
+                    if let Some(name) = name {
+                        // name is none only for the wrap_with_directory, which cannot really be
+                        // propagated up but still the parent_id is allowed to be None
+                        let previous = self
+                            .persisted_cids
+                            .entry(parent_id)
+                            .or_insert(collected)
+                            .insert(name, leaf);
+
+                        assert!(previous.is_none());
+                    }
+
+                    return Some(Ok(TreeNode {
+                        path: self.full_path.as_str(),
+                        cid: self.cid.as_ref().unwrap(),
+                        total_size: self.total_size,
+                        block: &self.block_buffer,
+                    }));
+                }
+            }
+        }
+        None
+    }
+}
+
+impl<'a> Iterator for PostOrderIterator<'a> {
+    type Item = Result<OwnedTreeNode, TreeConstructionFailed>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next_borrowed()
+            .map(|res| res.map(TreeNode::into_owned))
+    }
+}
+
+/// Borrowed representation of a node in the tree.
```

Comment:
It would be cool to instead have some "`Cow<'a, TreeNode>` on steroids" here :smiley:

Reply:
Cow cannot be applied here for Cid nor Box<[u8]>. Did think about this though.